src/aeso_syntax.erl

```diff
@@ -57,7 +57,8 @@
 -type pragma() :: {compiler, '==' | '<' | '>' | '=<' | '>=', compiler_version()}.
 
 -type letval()  :: {letval, ann(), pat(), expr()}.
--type letfun()  :: {letfun, ann(), id(), [pat()], type(), expr()}.
+-type letfun()  :: {letfun, ann(), id(), [pat()], type(), expr()}
+                 | {letfun, ann(), id(), [pat()], type(), expr(), expr()}.
```

Comment:
For the user facing syntax I think it would be better to have just one variant for simplicity. If there are no guards, then the expr could be `true` which can be optimized out later on. Also, I would rather opt for comma separated multiple guards -- this would ease implementing assigning values in guards like you can do in Haskell (but that's for another task). Then, no guards would be just an empty list here. @UlfNorell what do you think?

Reply:
Not introducing redundant forms in the AST is good, and comma separated guards make sense. Especially if we want to add support for what is called pattern guards in Haskell land later. (What you have here is referred to as just *guards*.)